lib/watcher-middleware.js

```diff
@@ -0,0 +1,56 @@
+'use strict';
+
+var url = require('url');
+var path = require('path');
+
+var setResponseHeaders = require('./utils/response-header');
+var errorHandler = require('./utils/error-handler');
+
+module.exports = function watcherMiddleware(watcher, options) {
+  options = options || {};
+
+  if (!options.hasOwnProperty('autoIndex')) {
+    // set autoIndex to be true if not provided
+    options.autoIndex = true;
+  }
+
+  return function(request, response, next) {
+    watcher.then(function(hash) {
+      var outputPath = path.normalize(hash.directory);
+
+      // set the x-broccoli header containing per request info used by the broccoli-middleware
+      var urlToBeServed = request.url;
+      var urlObj = url.parse(urlToBeServed);
+      var filename = path.join(outputPath, decodeURIComponent(urlObj.pathname));
+      var broccoliInfo = {
+        'url': urlToBeServed,
+        'outputPath': outputPath,
+        'filename': filename,
+      };
+      request.headers['x-broccoli'] = broccoliInfo;
+
+      var updatedFileName = setResponseHeaders(request, response, next, {
```

Comment:
> Ah sorry I did not make my point very clearly. I totally agree with all the headers: what I'm surprised about is that they're set in the watcher middleware and not the serve asset middleware.

I would disagree here since there is a usecase where this will not work as expected. The purpose of the watcher middleware was to set all the generic headers and the serve file asset middleware to actually just serve from disk. The reason being: Imagine usecase where someone wants to insert a middleware between the watcher and serve asset middleware in order to serve assets themselves. If the file headers are set by the serve asset only then the custom header inserted in between will need to duplicate the behavior of setting headers. IMO that is undesirable. The custom middleware or any downstream middleware should only overwrite the headers.